packages/create-component-with-subscriptions/src/createComponentWithSubscriptions.js

```diff
@@ -0,0 +1,222 @@
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import React from 'react';
+import invariant from 'fbjs/lib/invariant';
+
+// TODO The below Flow types don't work for `Props`
+export function createComponent<Props, Subscription, Value>(
+  config: {|
+    // Specifies the name of the subscribable property.
+    // The subscription value will be passed along using this same name.
+    // In the case of a functional component, it will be passed as a `prop` with this name.
+    // For a class component, it will be set in `state` with this name.
+    property: string,
+
+    // Synchronously get the value for the subscribed property.
+    // Return undefined if the subscribable value is undefined,
+    // Or does not support synchronous reading (e.g. native Promise).
+    getValue: (props: Props) => Value,
+
+    // Setup a subscription for the subscribable value in props.
+    // Due to the variety of change event types, subscribers should provide their own handlers.
+    // Those handlers should not attempt to update state though;
+    // They should call the valueChangedCallback() instead when a subscription changes.
+    // You may optionally return a subscription value to later unsubscribe (e.g. event handler).
+    subscribe: (
+      props: Props,
+      valueChangedCallback: (value: Value) => void,
+    ) => Subscription,
+
+    // Unsubsribe from the subscribable value in props.
+    // The subscription value returned from subscribe() is passed as the second parameter.
+    unsubscribe: (props: Props, subscription: Subscription) => void,
+  |},
+  Component: React$ComponentType<*>,
+): React$ComponentType<Props> {
+  invariant(Component != null, 'Invalid subscribable Component specified');
+  invariant(
+    typeof config.property === 'string' && config.property !== '',
+    'Subscribable config must specify a subscribable property',
+  );
+  invariant(
+    typeof config.getValue === 'function',
+    'Subscribable config must specify a getValue function',
+  );
+  invariant(
+    typeof config.subscribe === 'function',
+    'Subscribable config must specify a subscribe function',
+  );
+  invariant(
+    typeof config.unsubscribe === 'function',
+    'Subscribable config must specify a unsubscribe function',
+  );
+
+  const {getValue, property, subscribe, unsubscribe} = config;
+
+  // Unique state key name to avoid conflicts.
+  const stateWrapperKey = `____${property}`;
+
+  // If possible, extend the specified component to add subscriptions.
+  // This preserves ref compatibility and avoids the overhead of an extra fiber.
+  let BaseClass = (Component: any);
+  const prototype = (Component: any).prototype;
+
+  // If this is a functional component, use a HOC.
+  // Since functional components can't have refs, that isn't a problem.
+  // Class component lifecycles are required, so a class component is needed anyway.
+  if (
+    typeof prototype !== 'object' ||
```

Comment:
technically the prototype is allowed to be a function. you can extend a function.